apps/cowswap-frontend/src/common/pure/OrderProgressBarV2/container/RenderProgressTopSection.tsx

```diff
@@ -0,0 +1,232 @@
+import { useMemo } from 'react'
+
+import STEP_IMAGE_CANCELLED from '@cowprotocol/assets/cow-swap/progressbar-step-cancelled.svg'
+import STEP_IMAGE_EXPIRED from '@cowprotocol/assets/cow-swap/progressbar-step-expired.svg'
+import STEP_IMAGE_SOLVING from '@cowprotocol/assets/cow-swap/progressbar-step-solving.svg'
+import STEP_IMAGE_UNFILLABLE from '@cowprotocol/assets/cow-swap/progressbar-step-unfillable.svg'
+import STEP_LOTTIE_EXECUTING from '@cowprotocol/assets/lottie/progressbar-step-executing.json'
+import STEP_LOTTIE_NEXTBATCH from '@cowprotocol/assets/lottie/progressbar-step-nextbatch.json'
+import LOTTIE_TIME_EXPIRED_DARK from '@cowprotocol/assets/lottie/time-expired-dark.json'
+import { getRandomInt } from '@cowprotocol/common-utils'
+import { ProductLogo, ProductVariant, UI } from '@cowprotocol/ui'
+
+import { AnimatePresence, motion } from 'framer-motion'
+import Lottie from 'lottie-react'
+import SVG from 'react-inlinesvg'
+import { Textfit } from 'react-textfit'
+
+import * as styledEl from './styled'
+
+import { CHAIN_SPECIFIC_BENEFITS, SURPLUS_IMAGES } from '../constants'
+import { truncateWithEllipsis } from '../helpers'
+import AnimatedTokens from '../steps/AnimatedToken'
+import CircularCountdown from '../steps/CircularCountdown'
+import OrderIntent from '../steps/OrderIntent'
+import { OrderProgressBarV2Props } from '../types'
+
+function RenderProgressTopSection({
+  stepName,
+  order,
+  countdown,
+  chainId,
+  surplusData,
+  debugForceShowSurplus = false,
+}: Pick<OrderProgressBarV2Props, 'stepName' | 'order' | 'countdown' | 'chainId' | 'surplusData'> & {
+  debugForceShowSurplus?: boolean
+}) {
+  const hideIntent = stepName === 'finished' || stepName === 'cancellationFailed'
+
+  const { randomImage, randomBenefit } = useMemo(() => {
+    const benefits = CHAIN_SPECIFIC_BENEFITS[chainId]
+
+    return {
+      randomImage: SURPLUS_IMAGES[getRandomInt(0, SURPLUS_IMAGES.length - 1)],
+      randomBenefit: benefits[getRandomInt(0, benefits.length - 1)],
+    }
+  }, [chainId])
+
+  const { surplusPercent, showSurplus } = surplusData || {}
+  const shouldShowSurplus = debugForceShowSurplus || showSurplus
+  const surplusPercentValue = surplusPercent ? parseFloat(surplusPercent).toFixed(2) : 'N/A'
+
+  const content = useMemo(() => {
```

Comment:
Maybe this content also should be decomposed into several components (a component per case)